src/predicate_lang/predicate_lang.mli

```diff
@@ -22,44 +23,38 @@ val not_union : 'a t list -> 'a t
 
 val any : 'a t
 
-val decode_one : 'a Decoder.t -> 'a t Decoder.t
+val decode_one : 'a Decoder.t -> ('a -> 'b t) -> 'b t Decoder.t
 
-val decode : 'a Decoder.t -> 'a t Decoder.t
+val decode : 'a Decoder.t -> ('a -> 'b t) -> 'b t Decoder.t
 
-val encode : 'a Encoder.t -> 'a t Encoder.t
+val encode : 'a Encoder.t -> (Dune_glob.V1.t -> Dune_sexp.t) -> 'a t Encoder.t
```

Comment:
I think this type signature is not tight enough. The second argument should be parameterised by `'a` so that you only need to provide the function for the `'a = string` case. Or is it somehow not possible?

Reply:
Perhaps it's possible, but I'm not sure how to do it. There's no way constrain a constructor to be provided when `'a <> string`.

Reply:
Maybe something like this could work?

```ocaml
type 'a element = 
  | Literal : 'a -> 'a element
  | Glob : string -> string element

type 'a t = 
  | Element : 'a element -> 'a t
  | ...

val encode : ('a element -> 'a Encoder.t) -> 'a t Encoder.t
```
Then, when `'a = string`, you'll need to handle both constructors and provide a suitable encoder for globs, but when `'a = int`, you'll only need to provide one.

Reply:
At this point, I don't think we need the gadt at all. We can go back to the original representation:

```ocaml
type 'a t =
   | Element of 'a
   | ...
```

And then instantiate `'a t` with just:

```ocaml
type element = Glob of Glob.t | Literal of string
```

And use either `int Predicate_lang.t` or `element Predicate_lang.t` without any `assert false`.

The problem is that we lose the more "unboxed" representation that we wanted to have.

Reply:
Ah, of course, you're right.

Could we add a second type parameter? It would probably be a little ugly but maybe workable:

```ocaml
type has_no_globs = |
type has_globs = |

type ('a, 'b) t0 =
  | Element : 'a -> ('a, 'b) t
  | Glob : Dune_glob.V1.t -> (string, has_globs) t
  | Compl : ('a, 'b) t -> ('a, 'b) t
  ...

type 'a t = ('a, has_no_globs) t

let encode encoder = ...

module With_globs = struct
  type 'a t = ('a, has_globs) t

  let encode encoder glob_to_sexp = ...
end
```

Reply:
Right, that's a possibility. Or we could just not have a polymorphic encoding function at all and just have encoders for `string t` and `int` because that's all we use. Anyway, let's think it over tomorrow.